engine/src/main/java/org/hibernate/search/util/impl/FileHelper.java

```diff
@@ -84,58 +84,50 @@ else if ( !destination.isDirectory() ) {
 		}
 	}
 
-	public static void synchronize(File source, File destination, boolean smart) throws IOException {
+	public static void synchronize(Path source, Path destination, boolean smart) throws IOException {
 		synchronize( source, destination, smart, DEFAULT_COPY_BUFFER_SIZE );
 	}
 
-	public static void synchronize(File source, File destination, boolean smart, long chunkSize) throws IOException {
+	public static void synchronize(Path source, Path destination, boolean smart, long chunkSize) throws IOException {
 		if ( chunkSize <= 0 ) {
 			log.checkSizeMustBePositive();
 			chunkSize = DEFAULT_COPY_BUFFER_SIZE;
 		}
-		if ( source.isDirectory() ) {
-			if ( !destination.exists() ) {
-				if ( !destination.mkdirs() ) {
-					throw new IOException( "Could not create path " + destination );
-				}
-			}
-			else if ( !destination.isDirectory() ) {
-				throw new IOException(
-						"Source and Destination not of the same type:"
-								+ source.getCanonicalPath() + " , " + destination.getCanonicalPath()
-				);
-			}
-			String[] sources = source.list();
-			Set<String> srcNames = new HashSet<String>( Arrays.asList( sources ) );
-			String[] dests = destination.list();
+		if ( Files.isDirectory( source ) ) {
+			Files.createDirectories( destination );
+			Set<Path> sources = Files.list( source ).collect( Collectors.toSet() );
+			Set<String> sourceFilenames = sources.stream().map( v -> v.getFileName().toString() ).collect( Collectors.toSet() );
+			Set<Path> dests = Files.list( destination ).collect( Collectors.toSet() );
```

Comment:
I'm pretty sure you don't need this set. Just using `Files.list( destination ).map` instead of `dests.stream().map` when building `destFilenames` below would work equally well.

And while we're at it, I don't think you need to collect destination files at all. This would work just fine for deletion, and destination files are not used anywhere else:

```
			//delete files not present in source
			try ( DirectoryStream<Path> dests = Files.newDirectoryStream( source ) ) {
				for ( Path dest : dests ) {
					if ( !sourceFilenames.contains( dest.getFileName().toString() ) ) {
						delete( dest );
					}
				}
			}
```

You could also avoid collecting source file names and just resolve each dest file name against the source, but I'm not sure there's a benefit in that.